File: csharp/Simplify.CSharp/Simplify.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Simplify
{
    public static class Simplifier
    {
        internal static double SquareDistance(IPoint left, IPoint right)
        {
            var (leftX, leftY, leftZ) = left.Coordinates;
            var (rightX, rightY, rightZ) = right.Coordinates;

            return Math.Pow(leftX - rightX, 2) + Math.Pow(leftY - rightY, 2) + Math.Pow(leftZ - rightZ, 2);
        }

        internal static double SegmentSquareDistance(IPoint point, IPoint start, IPoint end)
        {
            var (x, y, z) = point.Coordinates;
            var (startX, startY, startZ) = start.Coordinates;
            var (endX, endY, endZ) = end.Coordinates;

            var deltaX = endX - startX;
            var deltaY = endY - startY;
            var deltaZ = endZ - startZ;

            double Result(double rx, double ry, double rz)
            {
                var dx = x - rx;
                var dy = y - ry;
                var dz = z - rz;
                return dx * dx + dy * dy + dz * dz;
            }

            if (deltaX == 0 && deltaY == 0 && deltaZ == 0)
            {
                return Result(startX, startY, startZ);
            }

            var t = ((x - startX) * deltaX + (y - startY) * deltaY + (z - startZ) * deltaZ) /
                    (deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ);

            return t > 1
                ? Result(endX, endY, endZ)
                : Result(startX + deltaX * t, startY + deltaY * t, startZ + deltaZ * t);
        }

        internal static IList<IPoint> SimplifyRadialDistance(IList<IPoint> points, double tolerance)
        {
            if (points.Count <= 2)
            {
                return points;
            }

            var previous = points.First();
            var newPoints = new List<IPoint> {previous};
            IPoint? selectedPoint = null;

            foreach (var point in points.Skip(1))
            {
                selectedPoint = point;

                if (!(SquareDistance(selectedPoint, previous) > tolerance))
                {
                    continue;
                }

                newPoints.Add(selectedPoint);
                previous = selectedPoint;
            }

            if (selectedPoint != null && previous != selectedPoint)
            {
                newPoints.Add(selectedPoint);
            }

            return newPoints;
        }

        internal static IList<IPoint> SimplifyDouglasPeucker(IList<IPoint> points, double tolerance)
        {
            if (points.Count <= 2)
            {
                return points;
            }

            var first = points.First();
            var last = points.Last();

            var (index, distance) = points
                .Select((point, i) => (
                    Index: i,
                    Distance: point == first || point == last ? -1 : SegmentSquareDistance(point, first, last)))
                .OrderByDescending(data => data.Distance)
                .First();

            if (distance > tolerance)
            {
                return SimplifyDouglasPeucker(points.Take(index + 1).ToList(), tolerance)
                    .SkipLast(1)
                    .Concat(SimplifyDouglasPeucker(points.Skip(index).ToList(), tolerance))
                    .ToList();
            }

            return new[] {first, last};
        }

        public static IEnumerable<IPoint> Simplify(IEnumerable<IPoint> points, double tolerance,
            bool highQuality = false)
        {
            var pointList = points.ToList();
            if (pointList.Count <= 2)
            {
                return pointList;
            }

            var squaredTolerance = tolerance * tolerance;

            return highQuality
                ? SimplifyDouglasPeucker(pointList, squaredTolerance)
                : SimplifyDouglasPeucker(SimplifyRadialDistance(pointList, squaredTolerance), squaredTolerance);
        }
    }
}

File: go/pkg/simplify/simplify.go
package simplify

import "math"

func squareDistance(left *MappablePoint, right *MappablePoint) float64 {
	leftX, leftY, leftZ := (*left).MapPoint()
	rightX, rightY, rightZ := (*right).MapPoint()

	return math.Pow(leftX-rightX, 2) + math.Pow(leftY-rightY, 2) + math.Pow(leftZ-rightZ, 2)
}

func segmentSquareDistance(point *MappablePoint, start *MappablePoint, end *MappablePoint) float64 {
	x, y, z := (*point).MapPoint()
	startX, startY, startZ := (*start).MapPoint()
	endX, endY, endZ := (*end).MapPoint()

	deltaX := endX - startX
	deltaY := endY - startY
	deltaZ := endZ - startZ

	result := func(rx float64, ry float64, rz float64) float64 {
		dx := x - rx
		dy := y - ry
		dz := z - rz
		return dx*dx + dy*dy + dz*dz
	}

	if deltaX == 0 && deltaY == 0 && deltaZ == 0 {
		return result(startX, startY, startZ)
	}

	t := ((x-startX)*deltaX + (y-startY)*deltaY + (z-startZ)*deltaZ) / (deltaX*deltaX + deltaY*deltaY + deltaZ*deltaZ)

	if t > 1 {
		return result(endX, endY, endZ)
	}

	return result(startX+deltaX*t, startY+deltaY*t, startZ+deltaZ*t)
}

func simplifyRadialDistance(points []MappablePoint, tolerance float64) []MappablePoint {
	if len(points) <= 2 {
		return points
	}

	previous := points[0]
	var newPoints = []MappablePoint{previous}
	var selectedPoint MappablePoint

	for _, point := range points[1:] {
		selectedPoint = point

		if squareDistance(&selectedPoint, &previous) < tolerance {
			continue
		}

		newPoints = append(newPoints, selectedPoint)
		previous = selectedPoint
	}

	if selectedPoint != nil && previous != selectedPoint {
		newPoints = append(newPoints, selectedPoint)
	}

	return newPoints
}

func simplifyDouglasPeucker(points []MappablePoint, tolerance float64) []MappablePoint {
	if len(points) <= 2 {
		return points
	}

	first := points[0]
	last := points[len(points)-1]

	var foundIndex int
	var bestDistance float64
	for index, point := range points {
		if point == first || point == last {
			continue
		}

		distance := segmentSquareDistance(&point, &first, &last)
		if distance >= bestDistance {
			bestDistance = distance
			foundIndex = index
		}
	}

	if bestDistance > tolerance {
		leftPart := points[:foundIndex+1]
		rightPart := points[foundIndex:]

		left := simplifyDouglasPeucker(leftPart, tolerance)
		leftR := left[:len(left)-1]

		right := simplifyDouglasPeucker(rightPart, tolerance)

		result := append(leftR, right...)
		return result
	}

	return []MappablePoint{first, last}
}

func Simplify(points []MappablePoint, tolerance float64, highQuality bool) []MappablePoint {
	if len(points) <= 2 {
		return points
	}

	squaredTolerance := tolerance * tolerance

	if highQuality {
		return simplifyDouglasPeucker(points, squaredTolerance)
	}

	return simplifyDouglasPeucker(simplifyRadialDistance(points, squaredTolerance), squaredTolerance)
}
